### DISCUSSION TITLE
Double Checking LUT Interpolation

### DISCUSSION THREAD

**Discussion by Trainmaster9977** [ScanMountGoat/Smush-Material-Research#45] [Q&A] [answered]
Hello! 

While I know you have some LUT interpolation code here [https://github.com/ScanMountGoat/Smush-LUT/blob/master/src/interp.rs](url) , I do want to double check that this is how smash does that interpolation, and want to ask where the code for said interpolation is within the code. 

Thanks!

**Comment by ScanMountGoat** [comment, marked as answer]
The interpolation is done by the GPU hardware and determined by the sampler settings. You can find the sampling algorithms in places like the vulkan specification, but it shouldn't be relevant for anything shader related. In shader code, you can just write `vec3 color = texture(my_3d_tex, vec3(x,y,z)).rgb;`. I would not expect the assembly to be much more complex since interpolation is a hardware feature on any modern GPU. For color grading LUTs in Ultimate, the values use linear interpolation in X, Y, and Z to avoid needing a large 256x256x256 texture.

**Reply by Trainmaster9977** [reply]
Ok, thank you. I just wanted to double check that any interpolation was linear, and as such, anything between two points of the LUT could be represented by an *ax+b* equation. 

Thanks!

**Reply by ScanMountGoat** [reply]
I would be careful with making simplifying assumptions like that. Trilinear interpolation can be defined as three separate linear interpolations in the X, Y, and Z axes. The final value is actually a mix of 8 values from the texture. It need not be the case that all RGB values in the LUT lie on a line when plotting the RGB values in 3D space. https://en.wikipedia.org/wiki/Trilinear_interpolation

What are you trying to do that requires converting the LUT into an equation?
